ext/mbstring/mbstring.c

```diff
@@ -2183,6 +2190,196 @@ PHP_FUNCTION(mb_output_handler)
+            Z_PARAM_LONG(split_length)
+            Z_PARAM_STR(encoding)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (split_length <= 0) {
+        php_error_docref(NULL, E_WARNING, "The length of each segment must be greater than zero");
+        RETURN_FALSE;
+    }
+
+    /* fill mbfl_string structure */
+    string.no_language = MBSTRG(language);
+    string.encoding = php_mb_get_encoding(encoding);
+    if (!string.encoding) {
+        RETURN_FALSE;
+    }
+
+    p = (char unsigned *)ZSTR_VAL(str); /* string cursor pointer */
+    string.val = (char unsigned *)p;
+    string.len = ZSTR_LEN(str);
+    last = string.val + string.len; /* last string char pointer */
+
+
+    /*
+     * +----------------------------------------------------------------------+
+     * |                           3 scenarios                                |
+     * +----------------------------------------------------------------------+
+     */
+    mbfl_encoding = string.encoding;
+
+    /*
+     * +----------------------------------------------------------------------+
+     * | first scenario:    1,2,4-bytes fixed width encodings (head part)     |
+     * +----------------------------------------------------------------------+
+     */
+    if (mbfl_encoding->flag & MBFL_ENCTYPE_SBCS) { /* 1 byte */
+        mb_len = string.len;
+        chunk_len = (size_t)split_length; /* chunk length in bytes */
+
+    } else if (mbfl_encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE )) { /* 2 bytes */
+        mb_len = string.len>>1; /* eq. string.len / 2 String length in chars */
+        chunk_len = split_length<<1; /* eq. split_length * 2 chunk length in bytes */
+
+    } else if (mbfl_encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE )) { /* 4 bytes */
+        mb_len = string.len>>2; /* eq. string.len / 4 */
+        chunk_len = split_length<<2; /* eq. split_length * 4 */
+    /*
+     * +----------------------------------------------------------------------+
+     * | second scenario: "variable width encodings with length table"        |
+     * +----------------------------------------------------------------------+
+     */
+    } else if (mbfl_encoding->mblen_table != NULL) {
+        char unsigned const *mbtab = mbfl_encoding->mblen_table;
+
+        /* assume that we have 2-bytes characters */
+        array_init_size(return_value, (string.len>>1 + split_length) / split_length);
+
+        while (p < last) { /* split cycle work until the cursor has reached the last byte */
+            char const *chunk_p = p; /* chunk first byte pointer */
+            chunk_len = 0; /* chunk length in bytes */
+            for (uint32_t char_count = 0; char_count < split_length; ++char_count) {
+                char unsigned const m = mbtab[*p]; /* single character length table */
+                chunk_len += m;
+                p += m;
+            }
+            if (p > last) chunk_len -= p - last; /* check if chunk is in bounds */
```

Comment:
This issue looks unresolved still.

Reply:
I think you wrong.
```
        while (p < last) { /* split cycle work until the cursor has reached the last byte */
            char const *chunk_p = p; /* chunk first byte pointer */
            chunk_len = 0; /* chunk length in bytes */
            for (zend_long char_count = 0; char_count < split_length; ++char_count) {
                char unsigned const m = mbtab[*p]; /* single character length table */
                chunk_len += m;
                p += m;
            }
            if (p > last) chunk_len -= p - last; /* check if chunk is in bounds */
            add_next_index_stringl(return_value, chunk_p, chunk_len);
        }
```
Out of the bounds risk exists only when dereferencing  pointer reference `p`, so we need to check it before writing chunk to the result array.

If `p` going out the bound here:
```
            for (zend_long char_count = 0; char_count < split_length; ++char_count) {
                char unsigned const m = mbtab[*p]; /* single character length table */
                chunk_len += m;
                p += m;
            }
```

it will be fixed here:
```
            if (p > last) chunk_len -= p - last; /* check if chunk is in bounds */
            add_next_index_stringl(return_value, chunk_p, chunk_len);
```

So it's ok i think.

Reply:
You are using `*p` inside the loop, that's also a dereference.

Reply:
@nikiC, possible i was wrong. Цill such a test be proof of my theory?

```
<?php
/*short string*/
$string = "очень короткая строка";

/* long split */
var_dump(mb_str_split($string, 12));
/* very ling split  */
var_dump(mb_str_split($string, 20));
```

Reply:
Something like `mb_str_split("foobar", 1024 * 1024);` will probably make the issue obvious. You can also try running under valgrind.

Reply:
![image](https://user-images.githubusercontent.com/32621719/52631387-f6c0b880-2ebe-11e9-9be0-82f908c633f6.png)

Reply:
Here is what valgrind says when running your example:

```
==19738== Use of uninitialised value of size 8
==19738==    at 0x4897B6: zif_mb_str_split (mbstring.c:2387)
==19738==    by 0x7859B8: ZEND_DO_ICALL_SPEC_RETVAL_USED_HANDLER (zend_vm_execute.h:694)
==19738==    by 0x7FFC3D: execute_ex (zend_vm_execute.h:61516)
==19738==    by 0x8061BE: zend_execute (zend_vm_execute.h:67944)
==19738==    by 0x717E30: zend_execute_scripts (zend.c:1639)
==19738==    by 0x67CDD8: php_execute_script (main.c:2635)
==19738==    by 0x808F81: do_cli (php_cli.c:992)
==19738==    by 0x80A0F7: main (php_cli.c:1384)
==19738== 
array(2) {
  [0]=>
  string(23) "очень коротк"
  [1]=>
  string(17) "ая строка"
}
==19738== Invalid read of size 1
==19738==    at 0x4897A6: zif_mb_str_split (mbstring.c:2387)
==19738==    by 0x7859B8: ZEND_DO_ICALL_SPEC_RETVAL_USED_HANDLER (zend_vm_execute.h:694)
==19738==    by 0x7FFC3D: execute_ex (zend_vm_execute.h:61516)
==19738==    by 0x8061BE: zend_execute (zend_vm_execute.h:67944)
==19738==    by 0x717E30: zend_execute_scripts (zend.c:1639)
==19738==    by 0x67CDD8: php_execute_script (main.c:2635)
==19738==    by 0x808F81: do_cli (php_cli.c:992)
==19738==    by 0x80A0F7: main (php_cli.c:1384)
==19738==  Address 0xa2214b8 is 0 bytes after a block of size 72 alloc'd
==19738==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==19738==    by 0x6DF658: __zend_malloc (zend_alloc.c:2926)
==19738==    by 0x6DE909: _emalloc (zend_alloc.c:2516)
==19738==    by 0x7547A5: zend_string_alloc (zend_string.h:132)
==19738==    by 0x754815: zend_string_init (zend_string.h:154)
==19738==    by 0x755612: zend_new_interned_string_request (zend_string.c:239)
==19738==    by 0x6E1DA0: zval_make_interned_string (zend_compile.c:472)
==19738==    by 0x6E1E1E: zend_insert_literal (zend_compile.c:484)
==19738==    by 0x6E1F51: zend_add_literal (zend_compile.c:505)
==19738==    by 0x6E549C: zend_emit_op (zend_compile.c:1793)
==19738==    by 0x6E7812: zend_compile_assign (zend_compile.c:2708)
==19738==    by 0x6F886C: zend_compile_expr (zend_compile.c:8281)
```

You can run this yourself using `USE_ZEND_ALLOC=0 valgrind sapi/cli/php script.php`.

The fact that something doesn't crash doesn't mean that it's correct. If you want to get an actual segfault, try something like `sapi/cli/php -r 'mb_str_split("foobar", 1024 * 1024 * 128);'`.

Reply:
I was wrong. Fixed.